types/jasmine/ts3.1/jasmine-tests.ts

```diff
@@ -0,0 +1,1358 @@
+        return {
+            compare: (actual: any, floor: number, ceiling: number): jasmine.CustomMatcherResult => {
+                const pass = actual >= floor && actual <= ceiling;
+                const message = `expected ${actual} to be within range ${floor}-${ceiling}`;
+                return { message, pass };
+            },
+
+            negativeCompare: (actual: any, floor: number, ceiling: number): jasmine.CustomMatcherResult => {
+                const pass = actual < floor && actual > ceiling;
+                const message = `expected ${actual} not to be within range ${floor}-${ceiling}`;
+                return { message, pass };
+            }
+        };
+    }
+};
+// add the custom matchers to interface jasmine.Matchers via TypeScript declaration merging
+// if your test files import or export anything, you'll want to use:
+// declare global {
+//     namespace jasmine {
+//         interface Matchers {
+//             ...
+//         }
+//     }
+// }
+declare namespace jasmine {
+    interface Matchers<T> {
+        toBeGoofy(expected?: Expected<T>): boolean;
+        toBeWithinRange(expected?: Expected<T>, floor?: number, ceiling?: number): boolean;
+    }
+}
+
+describe("Custom matcher: 'toBeGoofy'", () => {
+    beforeEach(() => {
+        jasmine.addMatchers(customMatchers);
+    });
+
+    it("is available on an expectation", () => {
+        expect({
+            hyuk: 'gawrsh'
+        }).toBeGoofy();
+    });
+
+    it("can take an 'expected' parameter", () => {
+        expect({
+            hyuk: 'gawrsh is fun'
+        }).toBeGoofy({ hyuk: ' is fun' });
+    });
+
+    it("can take many 'expected' parameters", () => {
+        expect(2).toBeWithinRange(1, 3);
+    });
+
+    it("can use the custom negativeCompare method", () => {
+        const matcher = customMatchers["toBeWithinRange"](jasmine.matchersUtil, []);
+        const result = matcher.negativeCompare!(1, 2, 3);
+
+        expect(result.pass).toBe(false);
+        expect(result.message).toBe("expected 1 not to be within range 2-3");
+    });
+
+    it("can be negated", () => {
+        expect({
+            hyuk: 'this is fun'
+        }).not.toBeGoofy();
+    });
+
+    it("has a proper message on failure", () => {
+        const actual = { hyuk: 'this is fun' };
+
+        const matcher = customMatchers["toBeGoofy"](jasmine.matchersUtil, []);
+        const result = matcher.compare(actual, null);
+
+        expect(result.pass).toBe(false);
+        expect(result.message).toBe(`Expected ${actual} to be goofy, but it was not very goofy`);
+    });
+});
+
+describe('better typed spys', () => {
+    describe('a typed spy', () => {
+        const spy = jasmine.createSpy('spy', (num: number, str: string) => {
+            return `${num} and ${str}`;
+        });
+        it('has a typed returnValue', () => {
+            // $ExpectType (val: string) => Spy<(num: number, str: string) => string>
```

Comment:
I could have done:

```ts
// $ExpectType string
let argType: Parameters<typeof spyObj.method.and.returnValue>[0];
```

and the same for below but I figured this may be simpler